Replace Hooks with EventDispatcher Part 2

Art4 opened the pull request:
This is a follow up PR of #14640 and replaces more calls of `Hook::callAll()` with the new EventDispatcher.

This PR replaces the call of `Hook::callAll()` with the EventDispatcher in 31 places and 17 files. (90 places in 48 files are left)

Art4 committed:
Add events for post_local, post_local_start and post_local_end

Art4 committed:
Replace hooks post_local_start, post_local and post_local_end with EventDispatcher

Art4 committed:
Implement home_init hook as Event

Art4 committed:
Merge branch 'develop' into replace-hooks-with-eventdispatcher

Art4 committed:
Fix cs

Art4 committed:
Implement photo_upload_form hook as Event

Art4 committed:
Implement route_collection hook as CollectRoutesEvent

Art4 committed:
Fix code style

annando commented:
Can you execute `bin/console dbstructure dumpsql > database.sql` to update the database configuration? This will make the database check be happy again.

Art4 commented:
Why should I do this in this PR/branch?

annando commented:
It could be done in any open PR. One could of course create a PR just for the pipeline. Except, I can't do it at the moment, because I currently develop on a folder that is not connected to any database.

Art4 commented:
> One could of course create a PR just for the pipeline.

I'll do that. 👍

Update: see #14807

Art4 committed:
Merge branch 'develop' into replace-hooks-with-eventdispatcher

Art4 committed:
Merge branch 'develop' into replace-hooks-with-eventdispatcher

Art4 committed:
Create event for network_to_name hook

MrPetovan commented:
And now the PHPMD check is going to hound us forever 😭

Art4 committed:
Fix code style

Art4 committed:
Merge branch 'develop' into replace-hooks-with-eventdispatcher

Art4 committed:
Create Event for jot_tool hook

Art4 committed:
Fix code style

Art4 committed:
create event for conversation_start hook

Art4 committed:
Fix code style

Art4 committed:
Create Event for display_item hook

Art4 committed:
Remove not needed hook call

Art4 committed:
create event for render_location hook

Art4 committed:
create event for item_photo_menu hook

Art4 committed:
Create event for oembed_fetch_url hook

Art4 committed:
Fix code style

Art4 committed:
Create event for page_info_data hook

Art4 committed:
create event for smilie hook

Art4 committed:
fix code style

Art4 committed:
Create events for bbcode hooks

Art4 committed:
fix code style

Art4 committed:
Fix smilies tests

Art4 committed:
Fix Mastodon Status tests

Art4 committed:
Merge branch 'develop' into replace-hooks-with-eventdispatcher

Art4 committed:
Fix Transmitter test

Art4 committed:
Fix code style

Art4 committed:
Create event for html2bbcode hook

Art4 committed:
Fix code style

Art4 commented:
I just stumbled upon the `contact_block_end` hook and noticed that it never really worked. It was introduced in 2019.03 with commit 1f98fdfda6a3e2b2f4f9065e690087a325678c5e.

The code looks like this: https://github.com/friendica/friendica/blob/2024.12-1/src/Content/Widget/ContactBlock.php#L110-L121

```php
		$o = Renderer::replaceMacros($tpl, [
			'$contacts' => $contacts_title,
			'$nickname' => $profile['nickname'],
			'$viewcontacts' => DI::l10n()->t('View Contacts'),
			'$micropro' => $micropro,
		]);

		$arr = ['contacts' => $contacts, 'output' => $o];

		Hook::callAll('contact_block_end', $arr);

		return $o;
```

As one can see `$o` is set into the `$arr` array (but not as reference) and than provided to the hook call. At the end only `$o` is returned. Because `$o` is not set as reference the value of `$o` never changed.

I would like to refactor this into a `HtmlFilterEvent`. In this case we would drop support for the provided `$contacts` context.

```diff
-		$arr = ['contacts' => $contacts, 'output' => $o];
-
-		Hook::callAll('contact_block_end', $arr);
+		$o = $eventDispatcher->dispatch(
+			new HtmlFilterEvent(HtmlFilterEvent::CONTACT_BLOCK_END, $o),
+		)->getHtml();

		return $o;
```

This would be a breaking change but until today no one has complained about that so I assume that no one has used this hook so far. 

What do you think?

Art4 committed:
Create Event for contact_block_end hook

Art4 commented:
I've refactored this into a `HtmlFilterEvent` and dropped support for the provided `$contacts` context. Update for the `CHANGLOG.md` from #14722:

```
### Changed

- **BREAKING**: The `contact_block_end` hook provides a HTML string instead of an array.

### Fixed

The `contact_block_end` hook has been fixed and can now change the content of the contact widget.
```

MrPetovan commented on src/Content/Widget/ContactBlock.php:
```diff
@@ -113,9 +112,11 @@ public static function getHTML(array $profile, int $visitor_uid = null): string
-		$arr = ['contacts' => $contacts, 'output' => $o];
+		$eventDispatcher = DI::eventDispatcher();
 
-		Hook::callAll('contact_block_end', $arr);
+		$o = $eventDispatcher->dispatch(
+			new HtmlFilterEvent(HtmlFilterEvent::CONTACT_BLOCK_END, $o),
+		)->getHtml();
```
I'm not sure we should change the structure of existing hooks since we don't know who might be using it as is. In particular, the list of contacts would be completely unavailable for addons implementing this hook. I'm not against adding a different hook with a simplified structure, but existing hooks should remain the same.

Art4 replied:
I'm fully with you. The point is that this hook never worked. And because nobody has complained about it I assume that it was never used.

Update: To be precise: It's hypothetical possible that someone has useed this hook read-only, but I don't think that this is the case.

MrPetovan replied:
Sounds good, please proceed.

Art4 committed:
Create Event for jot_networks hook

Art4 committed:
Create Events for protocol support hooks

Art4 committed:
Fix code style

Art4 commented:
I recommend reviewing and merging this PR before it gets too big. I replaced the call of `Hook::callAll()` with the EventDispatcher in 31 places and 17 files. (90 places in 48 files are left)

MrPetovan commented:
I'll get on it.

MrPetovan commented on src/Content/OEmbed.php:
```diff
@@ -274,12 +282,14 @@ private static function formatObject(\Friendica\Object\OEmbed $oembed, int $urii
-$test = Proxy::proxifyHtml($ret, $uriid);
+
+		// FIXME: Why is $test unused?
+		$test = Proxy::proxifyHtml($ret, $uriid);
```
Probably a remnant of a local debug routine with a logger or var_dump call. In any case, this method doesn't have side effects so this call is superfluous. If it turns out to be useful for debug, it can be fished out of the git history.
```suggestion
```

Art4 committed:
Update src/Content/OEmbed.php

Co-authored-by: Hypolite Petovan <hypolite@mrpetovan.com>
